[BUGFIX beta] Overhaul queryRecord

pangratz opened the pull request:
Currently the use case for `store.queryRecord` is not communicated very
well and there has been some confusion on what the expected server
response should look like (array, or an object).

There are several issues with the current code base:

- in general, if the serializer returns an array for the primary data
  returned by `normalizeQueryRecordResponse`, then `store.queryRecord`
  resolves with an array

- if rest serializer is used and an array is returned by the adapter,
  then the first entry of the array is used as primary data and
  `store.queryRecord` resolves with the first record

- if json-api serializer is used and the primary data is an array, then
  `store.queryRecord` resolves with an array of the primary records

- the API documentation for `queryRecord` is similar to `query` and
  doesn't indicate when this method should be used in contrast to
  `store.query` and getting the first record of the array

- an assertion for the payload returned by the adapter for `queryRecord`
  not being an empty object has been added in 2.4, which is a regression
  to the behavior of 2.3

-----------------------------------------------------------------------

This commit addresses the above issues and makes the following changes:

- add assertion that `store.queryRecord` never resolves with an array

- add deprecation warning for the rest-serializers' `queryRecord`, if an
  array is returned instead of a single record

- removes the assertion that the returned payload from the adapter for
  `queryRecord` is not an empty object

- add assertion within json-api serializer that the primary data of the
  normalized response is not an array

---

I am not so sure about the deprecation for the `rest-serializer`, since it looks like currently also `findRecord` works when an array is returned (see [this test](https://github.com/emberjs/data/blob/v2.4.3/tests/integration/adapter/rest-adapter-test.js#L94-L114)). But since `json-api-adapter` doesn't allow `data: [...]` responses for `queryRecord`, this seems like an inconsistency which is a little surprising IMHO `¯\_(ツ)_/¯` .

This addresses issues raised in #3977, #4227 and #4255.

sly7-7 commented on addon/-private/system/store.js:
````diff
@@ -935,15 +935,75 @@ Store = Service.extend({
+    The request is made through the adapters' `queryRecord`:
+
+    ```javascript
+    // app/adapters/application.js
````
maybe pointing to app/adapters/user.js would be better here ?

pangratz replied:
:+1: You're absolutely right!

sly7-7 commented on addon/-private/system/store.js:
````diff
@@ -935,15 +935,75 @@ Store = Service.extend({
+    });
+    ```
+
+    This method returns a promise, which is resolves with the found record.
````
`which is resolved` / or `which resolves`

pangratz replied:
lol, english, how do you even work.

sly7-7 commented on addon/-private/system/store/finders.js:
```diff
@@ -196,6 +196,11 @@ export function _queryRecord(adapter, store, typeClass, query) {
+      assert("Expected the primary data returned by the serializer for a `queryRecord` response to be a single object but instead it was an array.", !Array.isArray(payload.data), {
+        id: 'ds.store.queryRecord-array-response'
+      });
+
```
At first sight it seemed that was not necessary here, but I guess it handles the case of customs adapters/serializers, right ?

pangratz replied:
Yep :+1:

pangratz commented:
Thanks for the feedback @sly7-7!

sly7-7 commented:
I think this is a good answer to related issues, disambiguate, and making the things consistent, I'm all for it

pangratz commented:
This has been discussed in the team meeting and overall the taken approach looks good.

I will take a look at #3916 to see if it can be addressed with this PR as well.

bmac commented on addon/serializers/json-api.js:
```diff
@@ -199,6 +199,13 @@ const JSONAPISerializer = JSONSerializer.extend({
+      assert('Expected the primary data returned by the serializer for a `queryRecord` response to be a single object but instead it was an array.', false, {
+        id: 'ds.serializer.json-api.queryRecord-array-response'
+      });
+    }
```
What are your thoughts on doing this in `normalizeQueryRecordResponse`?

pangratz commented:
I've updated this PR to address the following issues:

- put assertion for `queryRecord` being an array in the `json-api` serializer from `_normalizeResponse` to `normalizeQueryRecordResponse`, which is a better place as pointed out by @bmac 👍 

- remove assertion for payload not being empty in the [`_queryRecord` method of `finder.js`](https://git.io/vw3WF). This assertion (introduced in #3916) caused a regression which has been reported in #4310. Removing the assertion allows to return an empty object again, when the `rest` adapter is used. Tests for `store.queryRecord` using the `rest` and `json-api` adapter have been added, so the expected behavior for those methods are tested and future regressions are caught.

- updated the documentation for `store.queryRecord` to point out the behavior: the returned promise resolves with either the found record or `null` when the adapter returned nothing for the primary data

taras commented:
Just to clarify the clarification, does this PR ensure that `queryRecord` always returns a single object and not an array?

pangratz commented:
> Just to clarify the clarification, does this PR ensure that queryRecord always returns a single object and not an array?

@taras yes, `store.queryRecord` [will always resolve with a single record](https://github.com/pangratz/data/blob/overhaul-queryRecord/addon/-private/system/store/finders.js#L199-L201). Is that what you meant?

taras commented:
Cool, thank you :)

bmac commented on addon/serializers/json-api.js:
```diff
@@ -202,6 +202,18 @@ const JSONAPISerializer = JSONSerializer.extend({
+    if (Array.isArray(normalized.data)) {
+      assert('Expected the primary data returned by the serializer for a `queryRecord` response to be a single object but instead it was an array.', false, {
+        id: 'ds.serializer.json-api.queryRecord-array-response'
+      });
```
Instead of wrapping this in an if block should we replace `false` with `!Array.isArray(normalized.data)` so it gets stripped correctly is a prod build?

pangratz replied:
Absolutely! Don't know why I chose the if here 😕

bmac commented:
Thanks @pangratz
